apps/champions/lib/champions/units.ex

```diff
@@ -1,28 +1,356 @@
-    if Units.get_selected_units(user_id) |> Enum.any?(&(&1.slot == slot)) do
-      {:error, :slot_occupied}
+    if Units.get_selected_units(user_id) |> Enum.any?(&(&1.slot == slot)),
+      do: {:error, :slot_occupied},
+      else: Units.select_unit(user_id, unit_id, slot)
+  end
+
+  @doc """
+  Sets a unit as unselected for a user. Clears the `slot` field.
+
+  Returns `:not_found` if unit doesn't exist or if it's now owned by the user.
+  Returns the unit's new state if succesful.
+  """
+  def unselect_unit(user_id, unit_id), do: Units.unselect_unit(user_id, unit_id)
+
+  @doc """
+  Levels up a user's unit and substracts the currency cost from the user.
+
+  Returns `{:error, :not_found}` if unit doesn't exist or if it's not owned by user.
+  Returns `{:error, :cant_afford}` if user cannot afford the cost.
+  Returns `{:error, :cant_level_up}`if unit cant level up due to tier restrictions.
+  Returns `{:ok, unit: %Unit{}, user_currency: %UserCurrency{}}` if succesful.
+  """
+  def level_up(user_id, unit_id) do
+    with {:unit, {:ok, unit}} <- {:unit, Units.get_unit(unit_id)},
+         {:unit_owned, true} <- {:unit_owned, unit.user_id == user_id},
+         {:can_level_up, true} <- {:can_level_up, can_level_up(unit)},
+         costs = calculate_level_up_cost(unit),
+         {:can_afford, true} <-
+           {:can_afford, Currencies.can_afford(user_id, costs)} do
+      result =
+        Multi.new()
+        |> Multi.run(:unit, fn _, _ -> Units.add_level(unit) end)
+        |> Multi.run(:user_currency, fn _, _ ->
+          add_user_currencies(unit, costs)
+        end)
+        |> Transaction.run()
+
+      case result do
+        {:error, reason} ->
+          {:error, reason}
+
+        {:error, _, _, _} ->
+          {:error, :transaction}
+
+        {:ok, %{unit: unit, user_currency: user_currency}} ->
+          {:ok, %{unit: unit, user_currency: user_currency}}
+      end
     else
-      Units.select_unit(user_id, unit_id, slot)
+      {:unit, {:error, :not_found}} -> {:error, :not_found}
+      {:unit_owned, false} -> {:error, :not_found}
+      {:can_level_up, false} -> {:error, :cant_level_up}
+      {:can_afford, false} -> {:error, :cant_afford}
     end
   end
 
-  def unselect_unit(user_id, unit_id) do
-    Units.unselect_unit(user_id, unit_id)
+  defp add_user_currencies(unit, costs) do
+    result =
+      Enum.map(costs, fn {currency_id, cost} ->
+        Currencies.add_currency(unit.user_id, currency_id, -cost)
+      end)
+
+    if Enum.all?(result, fn
+         {:ok, _} -> true
+         _ -> false
+       end) do
+      {:ok,
+       Enum.map(result, fn {_ok, user_currency} ->
+         UserCurrency.preload_currency(user_currency)
+       end)}
+    else
+      {:error, "failed"}
+    end
   end
+
+  @doc """
+  Returns whether a unit can level up. Level is blocked by tier.
+  """
+  def can_level_up(unit), do: can_level_up(unit.tier, unit.unit_level)
+  defp can_level_up(1, level) when level < 20, do: true
```

Comment:
This smells weird, why don't we use a config file? Like a json so you don't need to redeploy when adding a new value.

Reply:
Sure. Would we fetch the json on every can_level_up call? Unit level up is something that happens a lot among all users, that could result in some overheard (I don't know how costly it is to fetch it everytime). Are there alternatives to this?

Reply:
I have created the issue for this here! https://github.com/lambdaclass/champions_of_mirra/issues/189